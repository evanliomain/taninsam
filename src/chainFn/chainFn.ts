/**
 * @module chain
 */

import { ChainFn, Links, Result } from '../@types';
import { log } from '../.internal';

/**
 * Enchain any function.
 * To get the value, apply .value()
 * To transform the value with a function f, apply .chain(f)
 * @param f the first function to apply in the chain
 */
export function chainFn<T>(f: (e) => T): ChainFn<T> {
  return _chainByFunction<T>(e => ({ result: f(e), links: {} }), {});

  function _chainByFunction<T>(
    _f: (e, l?: Links) => Result<T>,
    links: Links,
    linkName?: string
  ): ChainFn<T> {
    return {
      chain: _chain,
      link: (newLinkName: string): ChainFn<T> =>
        _chainByFunction(_f, links, newLinkName),
      value: () => e => _f(e).result,
      log: (flag: string): ChainFn<T> =>
        _chain((x, closureLink) => {
          log(flag, x, closureLink);
          return x;
        })
    };

    function _chain(g) {
      return _chainByFunction((x, closureLink = {}) => {
        const { result, links: inneriteClosureLinks } = _f(x, closureLink);

        if (undefined === linkName) {
          return {
            result: g(result, inneriteClosureLinks),
            links: inneriteClosureLinks
          };
        }
        const newClosureLink = {
          ...inneriteClosureLinks,
          [linkName]: result
        };
        return { result: g(result, newClosureLink), links: newClosureLink };
      }, links);
    }
  }
}
