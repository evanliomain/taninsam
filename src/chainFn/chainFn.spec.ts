import { chainFn } from './chainFn';

describe('chainFn function', () => {
  beforeEach(() => {
    console.group = jest.fn();
    console.groupEnd = jest.fn();
    console.table = jest.fn();
    console.log = jest.fn();
  });

  test('by function', () => {
    const mapFn = jest.fn(array => array.map((x: number) => x + 1));
    const filterFn = jest.fn(array => array.filter((x: number) => 0 === x % 2));

    const chainResult = chainFn(mapFn)
      .chain(filterFn)
      .value();

    const result = chainResult([1, 2, 3, 4]);

    expect(mapFn.mock.calls.length).toBe(1);
    expect(filterFn.mock.calls.length).toBe(1);
    expect(mapFn.mock.calls[0][0]).toEqual([1, 2, 3, 4]);
    expect(filterFn.mock.calls[0][0]).toEqual([2, 3, 4, 5]);
    expect(result).toEqual([2, 4]);
  });

  test('by function with links', () => {
    const mapFn = jest.fn((array, _) => {
      return array.map((x: number) => x + 1);
    });
    const filterFn = jest.fn((array, _) => {
      return array.filter((x: number) => 0 === x % 2);
    });
    const doubleFn = jest.fn((array, _) => {
      return array.map((x: number) => 2 * x);
    });
    const printRatio = jest.fn((array, links) => {
      return `${array.length}/${links.total.length}`;
    });

    const chainResult = chainFn(mapFn)
      .link('total')
      .chain(filterFn)
      .chain(doubleFn)
      .chain(printRatio)
      .value()([1, 2, 3, 4]);

    expect(chainResult).toMatchSnapshot();
  });

  describe('log chainFn', () => {
    test('with array', () => {
      const mapFn = jest.fn((array, _) => {
        return array.map((x: number) => x + 1);
      });
      const filterFn = jest.fn((array, _) => {
        return array.filter((x: number) => 0 === x % 2);
      });
      const doubleFn = jest.fn((array, _) => {
        console.log('doubleFn', array);

        return array.map((x: number) => 2 * x);
      });
      const printRatio = jest.fn((array, links) => {
        return `${array.length}/${links.total.length}`;
      });

      const chainResult = chainFn(mapFn)
        .log('maped')
        .link('total')
        .chain(filterFn)
        .log('filtered')
        .chain(doubleFn)
        .chain(printRatio)
        .value()([1, 2, 3, 4]);

      expect(console.group).toBeCalled();
      expect(console.table).toBeCalled();
      expect(console.groupEnd).toBeCalled();
      expect(chainResult).toMatchSnapshot();
    });

    test('with object', () => {
      chainFn(x => x)
        .log('test')
        .value()('test');
      expect(console.group).toBeCalled();
      expect(console.log).toBeCalled();
      expect(console.groupEnd).toBeCalled();
    });
  });
});
