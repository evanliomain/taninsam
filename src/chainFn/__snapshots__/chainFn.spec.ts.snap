// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chainFn function by function with links 1`] = `"2/4"`;

exports[`chainFn function log chainFn with array 1`] = `"2/4"`;
